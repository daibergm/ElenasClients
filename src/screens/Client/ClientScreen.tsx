import React from 'react';
import { SafeAreaView, Text } from 'react-native';

// @Styles
import styles from './styles';

function ClientScreen() {
  return (
    <SafeAreaView style={styles.container}>
      <Text>Cliente</Text>
    </SafeAreaView>
  );
}

export default ClientScreen;
