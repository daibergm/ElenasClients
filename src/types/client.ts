export type Client = {
  id: number;
  firstName: string;
  lastName: string;
  cellphone: string;
};
