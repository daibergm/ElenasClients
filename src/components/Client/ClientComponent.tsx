import React from 'react';
import { View, ScrollView, Text } from 'react-native';
import { Formik } from 'formik';
import * as yup from 'yup';
import { isEmpty } from 'lodash';

// @Assets
import styles from './styles';

// @Types
import { Client } from '../../types/';

// @Components
import { Button, Input } from '../common';

// @Constants
import { TEST_IDS, ERRORS } from '../../constants/';

const validationSchema = yup.object().shape({
  firstName: yup.string().trim().required('Este campo es obligatorio.'),
  lastName: yup.string().trim().required('Este campo es obligatorio.'),
  cellphone: yup.string().trim().required('Este campo es obligatorio.'),
  address: yup.string().trim().required('Este campo es obligatorio.'),
});

interface Props {
  submitFunction: (values: Client) => void;
  loading: boolean;
  mutationLoading: boolean;
  apiError?: string;
  id?: number;
  client: Client;
}

function ClientComponent({
  submitFunction,
  loading,
  mutationLoading,
  id,
  client,
  apiError,
}: Props) {
  const onSubmit = (values: Client) => submitFunction(values);

  return (
    <View style={styles.container}>
      <Formik
        initialValues={client}
        validationSchema={validationSchema}
        onSubmit={onSubmit}
        enableReinitialize>
        {({
          values,
          handleChange,
          errors,
          setFieldTouched,
          isValid,
          handleSubmit,
          touched,
          dirty,
        }) => (
          <ScrollView
            alwaysBounceVertical={false}
            showsVerticalScrollIndicator={false}
            style={styles.scroll}
            keyboardShouldPersistTaps="handled">
            <View style={styles.subContainer}>
              <Input
                testID={TEST_IDS.firstNameInput}
                value={values.firstName}
                placeholder="Nombre"
                errorMessage={
                  touched.firstName ? (errors.firstName as string) : undefined
                }
                onChangeText={handleChange('firstName')}
                onBlur={() => setFieldTouched('firstName')}
              />
              <Input
                testID={TEST_IDS.lastNameInput}
                value={values.lastName}
                placeholder="Apellido"
                errorMessage={
                  touched.lastName ? (errors.lastName as string) : undefined
                }
                onChangeText={handleChange('lastName')}
                onBlur={() => setFieldTouched('lastName')}
              />
              <Input
                testID={TEST_IDS.cedulaInput}
                value={values.cedula}
                placeholder="Cédula"
                errorMessage={
                  touched.cedula ? (errors.cedula as string) : undefined
                }
                onChangeText={handleChange('cedula')}
                onBlur={() => setFieldTouched('cedula')}
              />
              <Input
                testID={TEST_IDS.emailInput}
                value={values.email}
                placeholder="Correo electrónico"
                errorMessage={
                  touched.email ? (errors.email as string) : undefined
                }
                onChangeText={handleChange('email')}
                onBlur={() => setFieldTouched('email')}
              />
              <Input
                testID={TEST_IDS.cellphoneInput}
                value={values.cellphone}
                placeholder="Celular"
                errorMessage={
                  touched.cellphone ? (errors.cellphone as string) : undefined
                }
                onChangeText={handleChange('cellphone')}
                onBlur={() => setFieldTouched('cellphone')}
              />
              <Input
                testID={TEST_IDS.addressInput}
                value={values.address}
                placeholder="Dirección"
                errorMessage={
                  touched.address ? (errors.address as string) : undefined
                }
                onChangeText={handleChange('address')}
                onBlur={() => setFieldTouched('address')}
              />
              <View style={styles.buttonContainer}>
                <Button
                  title={id ? 'EDITAR' : 'CREAR'}
                  testID={TEST_IDS.submitButton}
                  disabled={(!isValid || !dirty) && !id}
                  onPress={handleSubmit}
                  loading={loading || mutationLoading}
                />
              </View>
              {!isEmpty(apiError) && (
                <View>
                  <Text style={styles.errorLabel}>
                    {ERRORS[apiError as string]}
                  </Text>
                </View>
              )}
            </View>
          </ScrollView>
        )}
      </Formik>
    </View>
  );
}

export default ClientComponent;
