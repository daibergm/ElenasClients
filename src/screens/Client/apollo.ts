import { gql } from '@apollo/client';

// @Queries
export const CLIENT_QUERY = gql`
  query client($ids: [Int!]) {
    clientsSearch(ids: $ids) {
      ... on ClientPagination {
        results {
          id
          firstName
          lastName
          cedula
          email
          cellphone
          address
        }
      }
    }
  }
`;

// @Mutations
export const CREATE_MUTATION = gql`
  mutation createClient($input: ClientInput!) {
    createClient(input: $input) {
      ... on Client {
        id
      }
    }
  }
`;

export const UPDATE_MUTATION = gql`
  mutation updateClient($id: Int!, $input: ClientInput!) {
    updateClient(id: $id, input: $input) {
      ... on Client {
        id
      }
    }
  }
`;
