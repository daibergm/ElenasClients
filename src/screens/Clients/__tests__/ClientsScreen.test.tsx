import React from 'react';
import { render } from '@testing-library/react-native';
import { ApolloProvider } from '@apollo/client';

// @Screens
import Client from '../ClientsScreen';

// @Config
import apolloClient from '../../../config/apollo';

describe('Client test suits', () => {
  it('Render correctly', () => {
    const client = render(
      <ApolloProvider client={apolloClient}>
        <Client />
      </ApolloProvider>,
    );
    expect(client).toMatchSnapshot();
  });
});
