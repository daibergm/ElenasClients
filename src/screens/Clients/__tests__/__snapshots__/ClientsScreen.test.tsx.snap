// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Client test suits Render correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    ListEmptyComponent={[Function]}
    contentContainerStyle={
      Object {
        "flex": 1,
        "justifyContent": "center",
      }
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    showsVerticalScrollIndicator={false}
    stickyHeaderIndices={Array []}
    style={
      Object {
        "paddingHorizontal": 15,
        "paddingVertical": 25,
      }
    }
    viewabilityConfigCallbackPairs={Array []}
  >
    <View>
      <View
        style={
          Object {
            "alignItems": "center",
          }
        }
        testID="emptyList"
      >
        <Text
          style={
            Object {
              "fontSize": 15,
              "fontWeight": "600",
              "marginTop": 20,
            }
          }
        >
          Buscar Clientes
        </Text>
        <Text
          style={
            Object {
              "color": "grey",
              "fontSize": 15,
              "marginTop": 12,
            }
          }
        >
          Lo sentimos, no pudimos encontrar ning√∫n cliente.
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      Object {
        "backgroundColor": "#FFFFFF",
        "height": "100%",
        "justifyContent": "center",
        "position": "absolute",
        "width": "100%",
      }
    }
  >
    <ActivityIndicator
      color="#00adee"
      size="large"
    />
  </View>
</RCTSafeAreaView>
`;
